File: Lesson/Lesson 25 Matrix keyboard expansion experiment/Java/keyboard.java
import com.pi4j.wiringpi.Gpio;

public class keyboard {
    static int[] pin_val = new int[] {3, 4, 5, 6, 1, 24, 28, 29};
	static int[][] key = new int[][] {	{'1','2','3','A'},
										{'4','5','6','B'},
										{'7','8','9','C'},
										{'*','0','#','D'}	};
    static {
        // setup wiring pi
        if (Gpio.wiringPiSetup() == -1) {
            System.out.println(" ==>> Gpio SETUP FAILED");
        }
    }

	public static void out_set(int n) {
		for(int i = 0; i < 4; i++) {
			Gpio.pinMode(pin_val[i],Gpio.OUTPUT);
			Gpio.pinMode(pin_val[i+4], Gpio.INPUT);
		}
		if(n == 0) {
			Gpio.digitalWrite(pin_val[0],Gpio.HIGH);
			Gpio.digitalWrite(pin_val[1],Gpio.LOW);
			Gpio.digitalWrite(pin_val[2],Gpio.LOW);
			Gpio.digitalWrite(pin_val[3],Gpio.LOW);
		}
		if(n == 1) {
			Gpio.digitalWrite(pin_val[0],Gpio.LOW);
			Gpio.digitalWrite(pin_val[1],Gpio.HIGH);
			Gpio.digitalWrite(pin_val[2],Gpio.LOW);
			Gpio.digitalWrite(pin_val[3],Gpio.LOW);
		}			
		if(n == 2) {
			Gpio.digitalWrite(pin_val[0],Gpio.LOW);
			Gpio.digitalWrite(pin_val[1],Gpio.LOW);
			Gpio.digitalWrite(pin_val[2],Gpio.HIGH);
			Gpio.digitalWrite(pin_val[3],Gpio.LOW);
		}
		if(n == 3) {
			Gpio.digitalWrite(pin_val[0],Gpio.LOW);
			Gpio.digitalWrite(pin_val[1],Gpio.LOW);
			Gpio.digitalWrite(pin_val[2],Gpio.LOW);
			Gpio.digitalWrite(pin_val[3],Gpio.HIGH);
		}
	}

    public static void main(String args[]) {

        for ( ; ;) { 
			for(int i = 0; i < 4; i++) {
				out_set(i);
				for(int m = 0; m < 4; m++) {
					if(Gpio.digitalRead(pin_val[m+4]) == 1) {
						Gpio.delay(100);
						if(Gpio.digitalRead(pin_val[m+4]) == 1) {
							System.out.printf("KEY :%c\n",key[i][m]);
						}
					}
				}
			}
		}
	}
}

File: Lesson/Lesson 25 Matrix keyboard expansion experiment/Python/keyboard.py
import RPi.GPIO as GPIO
import time

pin_val = [22, 23, 24, 25, 18, 19, 20, 21]
key = [
  ['1','2','3','A'],
  ['4','5','6','B'],
  ['7','8','9','C'],
  ['*','0','#','D']
]
GPIO.cleanup()
GPIO.setmode(GPIO.BCM)

def out_set(n):
    for i in range(0,4):
        GPIO.setup(pin_val[i],GPIO.OUT)
        GPIO.setup(pin_val[i+4], GPIO.IN, pull_up_down=GPIO.PUD_UP)
    if n == 0:
        GPIO.output(pin_val[0],GPIO.HIGH)
        GPIO.output(pin_val[1],GPIO.LOW)
        GPIO.output(pin_val[2],GPIO.LOW)
        GPIO.output(pin_val[3],GPIO.LOW)
    if n == 1:
        GPIO.output(pin_val[0],GPIO.LOW)
        GPIO.output(pin_val[1],GPIO.HIGH)
        GPIO.output(pin_val[2],GPIO.LOW)
        GPIO.output(pin_val[3],GPIO.LOW)            
    if n == 2:
        GPIO.output(pin_val[0],GPIO.LOW)
        GPIO.output(pin_val[1],GPIO.LOW)
        GPIO.output(pin_val[2],GPIO.HIGH)
        GPIO.output(pin_val[3],GPIO.LOW)
    if n == 3:
        GPIO.output(pin_val[0],GPIO.LOW)
        GPIO.output(pin_val[1],GPIO.LOW)
        GPIO.output(pin_val[2],GPIO.LOW)
        GPIO.output(pin_val[3],GPIO.HIGH)
       
while True:
    for n in range(0,4):
        out_set(n)
        for m in range(0,4):
            if GPIO.input(pin_val[m+4]):
                time.sleep(0.1)
                if GPIO.input(pin_val[m+4]):
                    print ('You enter the  key:',key[n][m])